File: lib/go/internal/types/TType.go
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

package types

// TType represents a Telepact schema type node.
type TType interface {
	GetTypeParameterCount() int
	Validate(value any, typeParameters []*TTypeDeclaration, ctx *ValidateContext) []*ValidationFailure
	GenerateRandomValue(blueprintValue any, useBlueprintValue bool, typeParameters []*TTypeDeclaration, ctx *GenerateContext) any
	GetName(ctx *ValidateContext) string
}

File: lib/ts/src/internal/types/TType.ts
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

import { ValidationFailure } from '../validation/ValidationFailure.js';
import { GenerateContext } from '../generation/GenerateContext.js';
import { ValidateContext } from '../validation/ValidateContext.js';
import { TTypeDeclaration } from './TTypeDeclaration.js';

export abstract class TType {
    abstract getTypeParameterCount(): number;

    abstract validate(value: any, typeParameters: TTypeDeclaration[], ctx: ValidateContext): ValidationFailure[];

    abstract generateRandomValue(
        blueprintValue: any,
        useBlueprintValue: boolean,
        typeParameters: TTypeDeclaration[],
        ctx: GenerateContext,
    ): any;

    abstract getName(): string;
}
